pkg/brbct/brbct.go

```diff
@@ -0,0 +1,341 @@
+
+			return nil
+		} else {
+			return fmt.Errorf("received invalid chunk or already sent echo")
+		}
+	})
+
+	brbpbdsl.UponEchoMessageReceived(m, func(from t.NodeID, id int64, chunk []byte, rootHash []byte, proof *commonpb.MerklePath) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id)
+
+		fromId, err := strconv.ParseInt(from.Pb(), 10, 64)
+		if err != nil {
+			return err
+		}
+
+		dsl.MerkleProofVerifyRequest(m, mc.MerkleProofVerifier, rootHash, chunk, proof, &hashEchoMessageContext{id: id, fromId: fromId, chunk: chunk, rootHash: rootHash, proof: proof})
+		return nil
+	})
+
+	dsl.MerkleProofVerifyResult(m, func(result bool, context *hashEchoMessageContext) error {
+		if context.id <= lastId {
+			return nil
+		}
+		currentState := state[context.id]
+		if result {
+			updateAccumulator(
+				&currentState.echos,
+				&currentState.echoMessagesCount,
+				&currentState.echoMessagesAccumulator,
+				string(context.rootHash),
+				context.chunk,
+				context.fromId,
+				params,
+			)
+			return nil
+		} else {
+			return fmt.Errorf("received invalid echo")
+		}
+	})
+
+	brbpbdsl.UponReadyMessageReceived(m, func(from t.NodeID, id int64, rootHash []byte) error {
+		if id <= lastId {
+			return nil
+		}
+		initialize(&state, id)
+
+		fromId, err := strconv.ParseInt(from.Pb(), 10, 64)
+		if err != nil {
+			return err
+		}
+
+		stringRepl := string(rootHash)
+
+		currentState := state[id]
+
+		if _, ok := currentState.readyMessagesReceived[stringRepl]; !ok {
+			currentState.readyMessagesReceived[stringRepl] = make([]bool, params.GetN())
+		}
+		if !currentState.readyMessagesReceived[stringRepl][fromId] {
+			currentState.readyMessagesReceived[stringRepl][fromId] = true
+
+			currentState.readyMessagesCount[stringRepl]++
+			if currentState.readyMessagesCount[stringRepl] > currentState.readyMessagesAccumulator.count {
+				currentState.readyMessagesAccumulator.count = currentState.readyMessagesCount[stringRepl]
+				currentState.readyMessagesAccumulator.value = stringRepl
+			}
+		}
+
+		return nil
+	})
+
+	dsl.UponCondition(m, func() error {
+		for id, currentState := range state {
+			if id <= lastId {
+				continue
+			}
+			if (currentState.echoMessagesAccumulator.count >= params.GetN()-params.GetF()) && !currentState.sentReady {
+				currentState.sentReady = true
+				eventpbdsl.SendMessage(m, mc.Net, brbmsgs.ReadyMessage(mc.Self, id, []byte(currentState.echoMessagesAccumulator.value)), params.AllNodes)
+			}
+
+			if (currentState.readyMessagesAccumulator.count >= params.GetN()-2*params.GetF()) && !currentState.sentReady {
+				currentState.sentReady = true
+				eventpbdsl.SendMessage(m, mc.Net, brbmsgs.ReadyMessage(mc.Self, id, []byte(currentState.readyMessagesAccumulator.value)), params.AllNodes)
+			}
+
+			if currentState.readyMessagesAccumulator.count >= params.GetN()-params.GetF() && currentState.echoMessagesAccumulator.count > params.GetN()-2*params.GetF() && !currentState.delivered {
```

Comment:
There seems to be a subtle bug here: the message that got the maximum number of echoes may not be the same as the one that collected enough readys to be delivered. Also, I believe there should be `>=` instead of `>` for echoes.

```suggestion
			if currentState.readyMessagesAccumulator.count >= params.GetN()-params.GetF() && currentState.echoMessagesCount[currentState.readyMessagesAccumulator.value] >= params.GetN()-2*params.GetF() && !currentState.delivered {
```